demos/object_detection_demo_ssd_async/main.cpp

```diff
@@ -354,58 +401,197 @@ int main(int argc, char *argv[]) {
+                        modeInfo[currentMode].lastStartTime = switchTime;
+                    } else {
+                        presenter.handleKey(key);
+                    }
+                }
+            } else if (!emptyRequests.empty() && cap.isOpened()) {
+                auto startTime = std::chrono::steady_clock::now();
+                
+                cv::Mat frame;
+                if (!cap.read(frame)) {
+                    if (frame.empty()) {
+                        if (FLAGS_loop_input) {
+                            cap.open((FLAGS_i == "cam") ? 0 : FLAGS_i.c_str());
+                        } else cap.release();
+                        continue;
+                    } else {
+                        throw std::logic_error("Failed to get frame from cv::VideoCapture");
+                    }
+                }
 
-            if (isModeChanged) {
-                isModeChanged = false;
-            }
+                InferRequest::Ptr request;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
 
-            // Final point:
-            // in the truly Async mode we swap the NEXT and CURRENT requests for the next iteration
-            curr_frame = next_frame;
-            next_frame = cv::Mat();
-            if (isAsyncMode) {
-                async_infer_request_curr.swap(async_infer_request_next);
-            }
+                    request = std::move(emptyRequests.front());
+                    emptyRequests.pop_front();
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                ExecutionMode frameMode = currentMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                &outputName,
+                                                startTime,
+                                                frameMode,
+                                                &currentMode,
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outputName))),
+                                    startTime,
+                                    frameMode == currentMode
```

Comment:
It would be simpler to just save `frameMode` and compare it to `currentMode` in the main loop (when you're processing the completed request). This way, you won't have to capture `currentMode` in the callback, nor access it from multiple threads without synchronization.